### ISSUE TITLE
Compiler Frontend: Operators to enable(3)

### ISSUE THREAD

**Issue by seanshpark** [Samsung/ONE#1167]
Next step of #277 / #873 operators to enable.

Please read #277 (comment) for steps

**Comment by binarman** [comment]
@seanshpark 
What is `RELUN` operator?
I thought it is RELU_N1_TO_1 in #873  and #1032

**Comment by seanshpark** [comment]
> What is RELUN operator?

I think it's from pytorch that we cannot author from TensorFlow. I'll add a comment

**Comment by binarman** [comment]
@seanshpark 
Looks like DepthwiseConv2D already implemented, but not all attributes present in luci IR.
There are also no python example for this operator.

I suppose we need to add absent attributes and example, is it right?

**Comment by seanshpark** [comment]
> I suppose we need to add absent attributes and example, is it right?

@binarman , thanks for checking! Yes, we need to fill other codes :)

**Comment by seanshpark** [comment]
I think `DepthwiseConv2D ` came from `compiler/exo` when `luci` was first forked.

**Comment by ko-dimi** [comment]
@seanshpark, again I have suggestion on optimizing operator support process. Now we review draft with all commits, and after that we cut draft into PRs per component (AFAIK this is formal rule we have), this means that we can do all luci changes in one PR: luci/lang, luci/import, luci/service, luci/logex, luci/export, luci/tests. Do you have motivated objections against doing all luci changes in one PR?

**Comment by binarman** [comment]
I have no objections against this proposal, if other reviewers are ok with it.

**Comment by seanshpark** [comment]
> this means that we can do all luci changes in one PR

Already added my opinion about this: https://github.com/Samsung/ONE/issues/277#issuecomment-629866592

**Comment by binarman** [comment]
@seanshpark 
I experimented with operators from the list and have some suggestions.
If you want, take a look at code I experimented with: [examples.zip](https://github.com/Samsung/ONE/files/4702322/examples.zip)

**batch_normlization**
  I tried to convert operator [tf.nn.batch_normalization](https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization) and toco converts it to `mul` + `add` sequence. We support both these operators. 
  **Should we mark it as already enabled?**

**fused_batch_norm**
I did not find this operator in main TF API, but it exists in raw ops: [tf.raw_ops.FusedBatchNorm](https://www.tensorflow.org/api_docs/python/tf/raw_ops/FusedBatchNorm).
Looks like it is same as simple batch norm, but represented in TF graph as one operator.
Unfortunately toco crashed when I tried to convert this operator.
  **I don't see this operator in any of our target networks, maybe postpone it until we find example?**

**DilatedConvolution**
  Dilation is a part of conv 2d operator(it is not supported in current version of luci).
  **Is it all we need to implement for this task?**

**Flatten**
  toco converts it to reshape (already supported).
  **Should we mark it as already enabled?**

**instance_norm**
  @d-krylov @kvochko @AShedko and everyone who is interested
  Just want to note that this is circle specific operator, so we need to support import and other stuff that is not implemented yet.
  There are nothing to do with tflite related components.

**MatMul**
  I suppose this is about [tf.linalg.matmul](https://www.tensorflow.org/api_docs/python/tf/linalg/matmul) operator
  Some forms of Matmul can be converted to FullyConnected(already supported).
  But others are not. so matmul is a custom op. Example of such form is in attached archive.
  **Should we support MatMul custom op?**

**MaxPoolWithArgMax**
  I was looking at [tf.nn.max_pool_with_argmax](https://www.tensorflow.org/api_docs/python/tf/nn/max_pool_with_argmax)
  This operator is converted to custom operator.
  **Should we support MaxPoolWithArgMax custom op?**

**Permute**
  I didn't find Permute operator in core TF, but it is present in keras: [tf.keras.layers.Permute](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Permute)
  It is converted to Transpose, that is already supported.
  **Should we mark Permute as already enabled?**

**RealDiv**
  As @kvochko mentioned in #1650 this operator is converted in ordinary div.
  **Should we mark RealDiv as already enabled?**

**Comment by binarman** [comment]
I also want to clarify, what to do with MatMul, because we already have enabled BatchMatMul ([tf.raw_ops.BatchMatMul](https://www.tensorflow.org/api_docs/python/tf/raw_ops/BatchMatMul)), should we update it or create a new operator type that looks like [tf.raw_ops.MatMul](https://www.tensorflow.org/api_docs/python/tf/raw_ops/MatMul)?

**Comment by seanshpark** [comment]
@binarman  , I've added your questions(?) into separate issues for better discussion.

**Comment by seanshpark** [comment]
@binarman , I'll work on `instance_norm`

**Comment by binarman** [comment]
@seanshpark 
FYI DepthToSpace is functional, but dumpers are missing (tfl/circle dumpers and logex), I think we will finish them tomorrow, but don't hurry closing this issue =)

**Comment by seanshpark** [comment]
> but dumpers are missing 

Thanks! I've remove `done` mark. I'll add it again after these lands :)

**Comment by seanshpark** [comment]
@binarman , I've posted #1804 for InstanceNorm, it's about adding transformation pass.
Will there be any one interested with this?

**Comment by binarman** [comment]
@seanshpark 
Looks like everything is done here
